components/Modal/ManageSources.js

```diff
@@ -16,104 +29,484 @@ const messages = defineMessages({
   },
   closeButtonLabel: {
     defaultMessage: "Close"
+  },
+  sourcePath: {
+    defaultMessage: "Source path",
+    description: "The path or url to the source repository"
+  },
+  name: {
+    defaultMessage: "Name",
+    description: "Name of source"
+  },
+  type: {
+    defaultMessage: "Type",
+    description: "Type of source"
+  },
+  security: {
+    defaultMessage: "Security"
+  },
+  check_ssl: {
+    defaultMessage: "Check SSL certificate"
+  },
+  check_gpg: {
+    defaultMessage: "Check GPG key"
+  },
+  selectOne: {
+    defaultMessage: "Select one"
+  },
+  typeRepo: {
+    defaultMessage: "yum repository"
+  },
+  typeMirrorlist: {
+    defaultMessage: "mirrorlist"
+  },
+  typeMetalink: {
+    defaultMessage: "metalink"
+  },
+  add: {
+    defaultMessage: "Add Source"
+  },
+  save: {
+    defaultMessage: "Add Source"
+  },
+  update: {
+    defaultMessage: "Update Source"
+  },
+  cancel: {
+    defaultMessage: "Cancel"
   }
 });
 
 class ManageSources extends React.Component {
-  constructor() {
-    super();
+  constructor(props) {
+    super(props);
+    this.state = {
+      showModal: false
+    };
+    this.open = this.open.bind(this);
+    this.close = this.close.bind(this);
+  }
+
+  open() {
+    this.setState({ showModal: true });
+  }
+
+  close() {
+    this.setState({ showModal: false });
+  }
+
+  render() {
+    return (
+      <>
+        <a href="#" onClick={this.open}>
+          <FormattedMessage
+            defaultMessage="Manage Sources"
+            description="User action for displaying the list of source repositories"
+          />
+        </a>
+        {this.state.showModal && (
+          <ManageSourcesModal
+            manageSources={this.props.manageSources}
+            addSource={this.props.addModalManageSourcesEntry}
+            removeSource={this.props.removeModalManageSourcesEntry}
+            clearError={this.props.modalManageSourcesFailure}
+            close={this.close}
+            intl={this.props.intl}
+          />
+        )}
+      </>
+    );
+  }
+}
+
+class ManageSourcesModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      addEntry: false,
+      url: "",
+      name: "",
+      type: "",
+      check_ssl: false,
+      check_gpg: false,
+      warningDuplicateName: false,
+      warningDuplicateUrl: false,
+      editName: ""
+    };
+    this.handleShowForm = this.handleShowForm.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.handleValidateName = this.handleValidateName.bind(this);
+    this.handleValidateUrl = this.handleValidateUrl.bind(this);
+    this.handleEditSource = this.handleEditSource.bind(this);
+    this.handleSubmitSource = this.handleSubmitSource.bind(this);
+  }
+
+  componentDidUpdate(prevProps) {
```

Comment:
Shouldn't the Modal we reset any time it's closed? No matter if it failed or succeded? 
I mean hitting Cancel, should never remember your previous choices, no?

Reply:
Right, that's the behavior I would expect if I refactor the form to be its own component like we discussed earlier. But currently, the form is part of the modal, and the modal remains open on submit due to the weird nested modal behavior that I have now.